libs/x-content/src/main/java/org/elasticsearch/xcontent/internal/EmbeddedImplClassLoader.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.xcontent.internal;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.UncheckedIOException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.security.SecureClassLoader;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+/**
+ * A class loader that is responsible for loading implementation classes and resources embedded within an archive.
+ *
+ * <p> This loader facilitates a scenario whereby an API can embed its implementation and dependencies all within the same archive as the
+ * API itself. The archive can be put directly on the class path, where it's API classes are loadable by the application class loader, but
+ * the embedded implementation and dependencies are not. When locating a concrete provider, the API can create an instance of an
+ * EmbeddedImplClassLoader to locate and load the implementation.
+ *
+ * <p> The archive typically consists of two disjoint logically groups:
+ *  1. the top-level classes and resources,
+ *  2. the embedded classes and resources
+ *
+ * <p> The top-level classes and resources are typically loaded and located, respectively, by the parent of an EmbeddedImplClassLoader
+ * loader. The embedded classes and resources, are located by the parent loader as pure resources with a provider specific name prefix, and
+ * classes are defined by the EmbeddedImplClassLoader. The list of prefixes is determined by reading the entries in the MANIFEST.TXT.
+ *
+ * <p> For example, the structure of the archive named x-content:
+ * <pre>
+ *  /org/elasticsearch/xcontent/XContent.class
+ *  /IMPL-JARS/x-content/MANIFEST.txt - contains x-content-impl.jar, dep-1.jar, dep-2.jar
+ *  /IMPL-JARS/x-content/x-content-impl.jar/xxx
+ *  /IMPL-JARS/x-content/dep-1.jar/abc
+ *  /IMPL-JARS/x-content/dep-2.jar/xyz
+ * </pre>
+ */
+public final class EmbeddedImplClassLoader extends SecureClassLoader {
+
+    private static final String IMPL_PREFIX = "IMPL-JARS/";
+    private static final String MANIFEST_FILE = "/MANIFEST.TXT";
+
+    private final List<String> prefixes;
+    private final ClassLoader parent;
+
+    private static List<String> getProviderPrefixes(ClassLoader parent, String providerName) {
+        final String providerPrefix = IMPL_PREFIX + providerName;
+        final InputStream is = parent.getResourceAsStream(providerPrefix + MANIFEST_FILE);
```

Comment:
Do we strictly need this manifest? Can't we scan the classpath for directories at the given prefix and find them that way rather than having to statically generate this manifest?

Reply:
Not without a bunch of additional work. Resources from the classloader are consumed via a URL. All you can do is load that URL. The way that others can "list" a directory there is either if the directory is actually on disk (a file URL) or by  loading another jar reader for the underlying jar file and iterating the entries. The manifest is much cleaner.